### PR TITLE
Imu relative atittude

### PR DESCRIPTION
**Pull request by mallamacimj**
# Major Updates
- Created Quaternion interpolation method
- Unit tests for quaternion interpolation
- Created quaternionBasedComplimentaryFilter determines the orientation quaternion
- Created adaptiveAccelGain (updates the accel complimentary gain based on acceleration)
- Created some basic IMU unit tests

# Minor Updates
- Added FakeIMU for unit testing
- Updated names of measured values in IMU (measuredAcc -> accelerationVec, measuredGyro -> angularVelocity, measuredMag -> magField). IMU should only store the body frame measurements of these values. If a inertial frame version of these measurements is needed, the getter function should do the mapping using orientation
- Updated IMU datastring to record useful info (nothing derived except orientation)
- getter and setter methods for accel & mag complimentary gains
- Created a detailed readme explaining the math

# Notes
This PR does not fully complete the IMU. It has all the needed functionality but leaves out some nice to haves (such as euler angle conversions). This will be added in another PR at some point.

### CODE REVIEW

File: src/Sensors/IMU/IMU.cpp [hunk 1]
```diff
@@ -54,14 +66,98 @@ namespace mmfs
         sprintf(data, "Initial Magnetic Field (uT): %.2f,%.2f,%.2f\n", initialMagField.x(), initialMagField.y(), initialMagField.z());
         return data;
     }
-    void IMU::update()
+
+    void IMU::quaternionBasedComplimentaryFilter(double dt)
```
Comment (DrewBrandt) [inline src/Sensors/IMU/IMU.cpp:70]: What calls this function?

File: src/Sensors/IMU/IMU.cpp [hunk 2]
```diff
@@ -54,14 +66,98 @@ namespace mmfs
...
+        // 10. Combine with gryo/acc orientation to correct yaw
+        Quaternion q_BI = q_wa_BI * delta_q_mag;
+        q_BI.normalize();
+        //-------------------------------------------------//
+
+        orientation = q_BI.conjugate();
     }
 
-    bool IMU::begin(bool useBiasCorrection)
+    double IMU::adaptiveAccelGain(double alphaBar, double t_1, double t_2)
```
Comment (DrewBrandt) [inline src/Sensors/IMU/IMU.cpp:147]: What calls this function?

File: src/Math/Quaternion.h [hunk 3]
```diff
@@ -149,6 +149,7 @@ class Quaternion {
   // similarly for result.z().
   //
   Vector<3> toEuler() const {
+    // TODO what Euler is this?
```
Comment (DrewBrandt) [inline src/Math/Quaternion.h:152]: What do you mean?

File: src/Sensors/IMU/BNO055.cpp [hunk 4]
```diff
@@ -21,9 +21,11 @@ namespace mmfs
 
     void BNO055::read()
     {
-        measuredAcc = convertIMUtoMMFS(bno.getVector(Adafruit_BNO055::VECTOR_LINEARACCEL));
-        measuredGyro = convertIMUtoMMFS(bno.getVector(Adafruit_BNO055::VECTOR_GYROSCOPE));
-        measuredMag = convertIMUtoMMFS(bno.getVector(Adafruit_BNO055::VECTOR_MAGNETOMETER));
+        accelerationVec = convertIMUtoMMFS(bno.getVector(Adafruit_BNO055::VECTOR_LINEARACCEL));
```
Comment (DrewBrandt) [inline src/Sensors/IMU/BNO055.cpp:24]: Will these already be in the inertial frame from the BNO055? Should we endeavor to make sure they aren't?

File: src/Sensors/IMU/IMU.cpp [hunk 5]
```diff
@@ -1,51 +1,63 @@
 #include "IMU.h"
+#include <stdio.h>
```
Comment (DrewBrandt) [inline src/Sensors/IMU/IMU.cpp:2]: Do we need this include?

File: src/Sensors/IMU/IMU.cpp [hunk 6]
```diff
@@ -1,51 +1,63 @@
...
+    Vector<3> IMU::getAccelerationGlobal()
     {
-        return measuredGyro;
+        Quaternion accelBody = Quaternion(0, accelerationVec.x(), accelerationVec.y(), accelerationVec.z());
+        Quaternion accelInterial = orientation * accelBody * orientation.conjugate();
+        return Vector<3>(accelInterial.x(), accelInterial.y(), accelInterial.z());
     }
 
-    Vector<3> IMU::getAccReading()
+    Vector<3> convertToEuler(const Quaternion &orientation)
```
Comment (DrewBrandt) [inline src/Sensors/IMU/IMU.cpp:44]: Is this a repeat of `Quaternion.h`'s method?

File: src/Sensors/IMU/IMU.h [hunk 7]
```diff
@@ -9,26 +9,21 @@
...
+        virtual Vector<3> getMagField();
+        virtual Vector<3> getAcceleration();
         virtual Vector<3> getAccelerationGlobal();
-        virtual Vector<3> getMagnetometerReading();
-        virtual Vector<3> getGyroReading();
-        virtual Vector<3> getAccReading();
+        virtual double getAccelBestFilteringAtStatic() {return accel_best_filtering_at_static;};
+        virtual void setAccelBestFilteringAtStatic(double a) {accel_best_filtering_at_static = a;};
+        virtual double getMagBestFilteringAtStatic() {return mag_best_filtering_at_static;};
+        virtual void setMagBestFilteringAtStatic(double m) {mag_best_filtering_at_static = m;};
```
Comment (DrewBrandt) [inline src/Sensors/IMU/IMU.h:26]: What are these functions for? Are they intended to be public?

We may run into multiple definition issues with this and need to move these into the .cpp. Probably ok for now though.

File: src/Sensors/IMU/README.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+## Absolute Attitude Determination
+
+### Quaternion Based Complementary Filter (w/ Adaptive Gain)
+
+Sources: 
+- https://ahrs.readthedocs.io/en/latest/filters/aqua.html#quaternion-based-complementary-filter
+- https://www.mdpi.com/1424-8220/15/8/19302
+
+Given gyroscope (angular velocity), accelermater (linear acceleration), magnatometer (magnetic field) measurements (and delta t), determine and propogate an absolute orientation quaternion.
```
Comment (DrewBrandt) [inline src/Sensors/IMU/README.md:64]: accelermater and magnatometer

File: src/Sensors/IMU/README.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+1. Determine rate of change of quaternion
+
+![Equation 2](https://latex.codecogs.com/svg.image?%7BB%5Catop%7D%5Cmathbf%7B%5Comega%7D_%7Bq,t_%7Bk%7D%7D=%5Cbegin%7Bbmatrix%7D0%5C%5C%5Comega_x%5C%5C%5Comega_y%5C%5C%5Comega_z%5Cend%7Bbmatrix%7D)
+
+${B \atop I}q_{w,t_k} = {I \atop B}q_w^{-1}$
+
+The above step is necessary as the quaternion we track in the code is the mapping from the body frame to the interial frame.
+
+${B \atop I}\dot{q}_{w,t_k} = -\frac{1}{2} {B \atop }\mathbf{\omega}_{q,t_{k}} {B \atop I}q_{w,t_k}$
```
Comment (DrewBrandt) [inline src/Sensors/IMU/README.md:80]: This line doesn't render

File: src/Sensors/IMU/images/equation (1).svg
Comment (DrewBrandt) [inline src/Sensors/IMU/images/equation (1).svg:1]: Would a gray text work better? idk how easy it is for you to change

File: test/test_imu/getExpectedResults.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
+import numpy as np
```
Comment (DrewBrandt) [inline test/test_imu/getExpectedResults.py:1]: What does this file do?

File: test/test_quaternion/getExpectedResults.py [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import numpy as np
```
Comment (DrewBrandt) [inline test/test_quaternion/getExpectedResults.py:1]: What does this file do?

File: src/Sensors/IMU/IMU.cpp (hunk 1 above)
Comment (mallamacimj) [reply src/Sensors/IMU/IMU.cpp:70]: This function can be thought of as util functions native to IMUs and get called during the hardware read() step if that hardware needs it to find orientation.

File: src/Sensors/IMU/IMU.cpp (hunk 2 above)
Comment (mallamacimj) [reply src/Sensors/IMU/IMU.cpp:147]: This function can be thought of as util functions native to IMUs and get called during the hardware read() step if that hardware needs it to find orientation.

For this specifically it is called by quaternionBasedComplimentaryFilter as well

File: src/Sensors/IMU/BNO055.cpp (hunk 4 above)
Comment (mallamacimj) [reply src/Sensors/IMU/BNO055.cpp:24]: No. The bno returns it's measurements in the body frame AND it returns an orientation quaternion. It is up to the user of the bno to do the transformation from the body to inertial using that orientation. 

This is how we should and I did set up our IMU class. We just add a helper function that does that transformation for you if you want.

File: src/Sensors/IMU/IMU.cpp (hunk 5 above)
Comment (mallamacimj) [reply src/Sensors/IMU/IMU.cpp:2]: Nope. I was print statement debugging and forgot to remove.

File: src/Sensors/IMU/IMU.h (hunk 7 above)
Comment (mallamacimj) [reply src/Sensors/IMU/IMU.h:26]: These are used to set filter constants used by the IMU and specifically the quaternionBasedComplimentaryFilter function. I will defer to you on where this should structurally live but they just need to be a getter and setter function for a double.

File: src/Sensors/IMU/IMU.cpp (hunk 6 above)
Comment (mallamacimj) [reply src/Sensors/IMU/IMU.cpp:42]: yes. it was a relic of old code. good catch

File: src/Sensors/IMU/images/equation (1).svg
Comment (mallamacimj) [reply src/Sensors/IMU/images/equation (1).svg:1]: probably... I'll make an issue and do it at some point

File: test/test_imu/getExpectedResults.py (hunk 10 above)
Comment (mallamacimj) [reply test/test_imu/getExpectedResults.py:1]: In my TestIMU, I give the filter some inputs and I want to validate the outputs. I am not going to sit here and do hand calcs for an hour to get the wrong answer so it just uses a python library that has implemented the filter to find the expected results to validate my implementation was correct.

File: test/test_quaternion/getExpectedResults.py (hunk 11 above)
Comment (mallamacimj) [reply test/test_quaternion/getExpectedResults.py:1]: In my QuaternionIMU, I give the filter some inputs and I want to validate the outputs. I am not going to sit here and do hand calcs for an hour to get the wrong answer so it just uses a python library that has implemented the filter to find the expected results to validate my implementation was correct.

**Comment by DrewBrandt** [comment]
Closes #72 now.

**Review by DrewBrandt — APPROVED** [review]

File: src/Sensors/IMU/IMU.cpp [hunk 12]
```diff
@@ -54,14 +55,98 @@ namespace mmfs
...
+        // and current orientation and propgated the orientation to the next time step.
+
+        //----------------- GYRO ORIENTATION --------------//
+        // 1. Determine rate of change of quaternion
+        Quaternion w_b = Quaternion{0, angularVelocity};
+        Quaternion q_w_BI = orientation.conjugate(); // Map from Inertia to body r_B = q_w_BI * r_I * q_w_BI^-1
+        Quaternion q_w_BI_dot = w_b * q_w_BI * -.5;
+
+        // 2. Propogate gyro orientation
+        q_w_BI = q_w_BI + q_w_BI_dot * dt;
```
[the remark this answers is not in the document]
Comment (mallamacimj) [reply src/Sensors/IMU/IMU.cpp:114]: this is exactly an Euler approximation and most definitely drifts over time. It'll get updated with info from the accel and mag vectors to hopefully slow that drift down considerably

**Pull request merged**